README.md
# The-Great-Agenda
An agenda app
who knows, could be useful to a person out there


Changes.txt
-Extension now uses 'chrome.storage' instead of the bland of 'localStorage' api


background.js
var hours = [7, 8, 9, 11, 13, 15, 17, 19, 21, 23];
//Fill an array with generated dates that have the day of today but have the hour of the hours desired to potentially ping someone.
var generatedHdates = [];
for (i = 0; i < hours.length; i++) {
  var generatedDate = new Date(getNow().year, getNow().month, getNow().day, hours[i]);
  generatedHdates.push(generatedDate);
}

function clr(id) {
  chrome.notifications.clear(id);
}
function getNow() {
  var rightNow = new Date();
  var now = {
    'total': rightNow,
    'year': rightNow.getFullYear(),
    'month': rightNow.getMonth(),
    'day': rightNow.getDate(),
    'hour': rightNow.getHours(),
    'minute': rightNow.getMinutes(),
    'second': rightNow.getSeconds(),
    'ms': rightNow.getMilliseconds()
  }

  return now;
}
function DiffofDatesInms(a, b) {
  //A and B are the dates a and b in ms
  var A = a.getTime();
  var B = b.getTime();

  return Math.abs(B - A);

}
function compareDates(a, b) {
  var A = a.getTime();
  var B = b.getTime();

  if (A < B) {
    return 'less';
  }
  if (A > B) {
    return 'greater';
  }
  if (A == B) {
    return 'equal';
  }
}
function getNextN() {
  var nextDate = 0;
  if (getNow().hour < 24) {
    for (i = generatedHdates.length-1; i >= 0; i = i - 1) {
      if (compareDates(generatedHdates[i], getNow().total) == 'greater') {
        nextDate = generatedHdates[i];
      }
      else {
        return nextDate;
      }
    }
  }
  else {
    return generatedHdates[0].setDate(generatedHdates[0].getDate() + 1);
    console.log("Setting to tmrw morning.");
  }
}
function setAtoNext() {
  if (getNow().hour < 24) {
    chrome.alarms.clearAll();

    console.log("Now: " + getNow().total + "(epoch: " + Date.now() + ")");
    console.log("Next notification: " + getNextN());

    try {
     var finalDate = getNextN().getTime();
    }
    catch {
      console.log("Next notification set to morning.");
      var morningCopy = generatedHdates[0]
      var finalDate = morningCopy.setDate(getNow().total.getDate() + 1);
      console.log("Morning notification, setting to: " + finalDate);
    }
     console.log('Timestamp set: ' + finalDate);
     chrome.alarms.create('ping', {when: finalDate});
  }
}
setAtoNext(); //on boot set A
var html = localStorage.getItem("tasklist");
chrome.runtime.onInstalled.addListener(function() {
  localStorage.setItem("lastAtime", getNow().total.toString())
  // chrome.storage.sync.set({"lastAtime": getNow().total.toString()});
  // var inst = {
  //   type: "basic",
  //   title: "Installed",
  //   message: 'Extension install',
  //   iconUrl: "chrome-extension://paomcbcgpoikdcjhbanhllhdbemcjokf/Agenda_32.png",
  //   buttons: [
  //     {
  //     title: 'OK'
  //     }
  //   ]
  // };
  // clr('install');
  // chrome.notifications.create('install', inst);
});

//7200000 = 2 hours in ms


function checkIfDueSoon() {
  function dateDiffInDays(a, b) {
    // Discard the time and time-zone information.
    const utc1 = Date.UTC(a.getFullYear(), a.getMonth(), a.getDate());
    const utc2 = Date.UTC(b.getFullYear(), b.getMonth(), b.getDate());

    return Math.floor((utc2 - utc1) / msPerDay);
  }
  var html = localStorage.getItem("tasklist");
  console.log(html);
  var div = document.createElement("div");
  div.innerHTML = html;
  var msPerDay = 1000 * 60 * 60 * 24;
  var tills = div.getElementsByClassName("daysTill");
  var items = div.getElementsByClassName("item");
  var dates = div.getElementsByClassName("dateStr");
  var len = tills.length;
  console.log(len);
  var sendNotfication;
  for (var i = 0; i < len; i += 1) {
    var today = new Date();
    var originalID = dates[i].id;
    var dateInfo = originalID.replace(/ok/g, ' ');
    var storedDate = new Date(dateInfo);
    var remainingDays = dateDiffInDays(today, storedDate);
    var nameOfAssign = items[i].parentElement.id;
    console.log(nameOfAssign);
    if (remainingDays <= 1) {
      sendNotfication = true;
    }
    else {
      sendNotfication = false;
    }

    if (remainingDays == 1) {
      var message = nameOfAssign + " is due tomorrow."
      var dueSoonopt = {
        type: "basic",
        title: "Assignment due",
        message: message,
        iconUrl: "chrome-extension://paomcbcgpoikdcjhbanhllhdbemcjokf/Agenda_32.png",
        buttons: [
          {
          title: 'Mark as done'
          }
        ]
      }
    }

    if (remainingDays == 0) {
      var message = nameOfAssign + " is due today."
      var dueSoonopt = {
        type: "basic",
        title: "Assignment due today",
        message: message,
        iconUrl: "chrome-extension://paomcbcgpoikdcjhbanhllhdbemcjokf/Agenda_32.png",
        buttons: [
          {
          title: 'Mark as done'
          }
        ]
      }
    }

    if (remainingDays < 0) {
      var message = nameOfAssign + " is overdue!"
      var dueSoonopt = {
        type: "basic",
        title: "Assignment overdue",
        message: message,
        iconUrl: "chrome-extension://paomcbcgpoikdcjhbanhllhdbemcjokf/Agenda_32.png",
        buttons: [
          {
          title: 'Mark as done'
          }
        ]
      }
    }

    if (sendNotfication) {
      clr(nameOfAssign);
      chrome.notifications.create(nameOfAssign, dueSoonopt);
      sendNotfication = false;
    }
  }
}
  checkIfDueSoon();
  var lastA = localStorage.getItem("lastAtime");
  console.log("Last alarm was " + DiffofDatesInms(getNow().total, new Date(lastA)) / 60000 + " minutes ago.");
  if (DiffofDatesInms(getNow().total, new Date(lastA)) > 7200000) {
    console.log("MISSED!");
    checkIfDueSoon();
  }

  chrome.alarms.onAlarm.addListener(function(alarm) {
    console.log("Beep from " + alarm.name);
    //Set last sucessful alarm time to right now.
    // chrome.storage.sync.set({"lastAtime": getNow().total.toString()});
    localStorage.setItem("lastAtime", getNow().total.toString());
    checkIfDueSoon();
    setAtoNext();
  });


chrome.notifications.onButtonClicked.addListener(function(notifId, btnIdx) {
    var htmlCopy = localStorage.getItem("tasklist");
    console.log(htmlCopy);
    var doc = new DOMParser().parseFromString(htmlCopy, "text/html");
    doc.innerHTML = htmlCopy;
    console.log(doc);
    console.log(notifId);
    //
    // var win = window.open('mainWindow.html');

    var targets = Array.prototype.slice.call(doc.getElementsByTagName("li"));
    console.log(targets);
    var finalindex = 0;
    var finalTarget = 0;
    targets.forEach((item, index) => {
      console.log(item.id);
      console.log(notifId);
      if (item.id.trim() == notifId.trim()) {
        console.log("match found.")
        console.log(targets[index]);
        finalTarget = targets[index].outerHTML;
      }
    });

    console.log(finalTarget);

     var removedTargetHtml = htmlCopy.replace(finalTarget, "");
     console.log(removedTargetHtml);
     localStorage.setItem("tasklist", removedTargetHtml);
     window.location.href="mainWindow.html";
     // chrome.tabs.create({
     //   active: true,
     //   url:  'mainWindow.html'
     // }, null);

    // chrome.storage.sync.set({'tasklist': removedTargetHtml});
    // win.close();
    //setTimeout(function(){target.remove(); chrome.storage.sync.set({'tasklist': list.innerHTML});}, 250);
});

document.addEventListener('DOMContentLoaded', () => {
  chrome.alarms.getAll(function(array) {
    console.log("ALARMS: ");
    console.log(array);
  });
  //This function checks if there are any assignments due soon. Decodes the date stored in the id of the list html elements.
});


// //a and b are two date objects
// function DiffofDatesInms(a, b) {
//   //A and B are the dates a and b in ms
//   var A = a.getTime();
//   var B = b.getTime();
//
//   return Math.abs(B - A);
//
// }


js/window.js
window.onload=function() {
  document.getElementById("Finish").click();
  document.getElementById("Link").focus();
  chrome.tabs.query({active: true, lastFocusedWindow: true}, tabs => {
    var url = tabs[0].url;
    document.getElementById("Link").value = url;
  });
};

//Get month, day, and year given a date in input form (ex: 2005-9-12)
//Acess with .month, .day, and .year properties of returned object
function getMDY(d) {
  var month = parseInt(d.slice(5,7));
  var day = parseInt(d.slice(8,10));
  var year = parseInt(d.slice(0,4));

  var finaldate = {
    'month': month,
    'day': day,
    'year': year
  }

  return finaldate;
}

function isUrl(string) {
  try {
    new URL(string);
  } catch (_) {
    return false;
  }

  return true;
}

var months = {
  1: "January",
  2: "February",
  3: "March",
  4: "April",
  5: "May",
  6: "June",
  7: "July",
  8: "August",
  9: "September",
  10: "October",
  11: "November",
  12: "December"
};

function changePrettyDate() {
  var prettyDate = document.getElementById("prettyDate");
  var date = document.getElementById("Date").value;
  var d = new Date(document.getElementById("Date").value);
  date = getMDY(date);
  var dateString = months[date.month] + " " + date.day + " ";
  prettyDate.innerHTML = dateString;
}

function changeTitle() {
  chrome.tabs.query({active: true, lastFocusedWindow: true}, tabs => {
    var title = tabs[0].title;
    var url = tabs[0].url;
    document.getElementById("LinkLabel").innerHTML = title;
    var favicon = "https://s2.googleusercontent.com/s2/favicons?domain_url=" + url
    document.getElementById("FAVICON").setAttribute("src", favicon);
  });
}
const _MS_PER_DAY = 1000 * 60 * 60 * 24;
// a and b are javascript Date objects
function dateDiffInDays(a, b) {
  // Discard the time and time-zone information.
  const utc1 = Date.UTC(a.getFullYear(), a.getMonth(), a.getDate());
  const utc2 = Date.UTC(b.getFullYear(), b.getMonth(), b.getDate());

  return Math.floor((utc2 - utc1) / _MS_PER_DAY);
}

document.addEventListener("DOMContentLoaded", function(event) {
  changeTitle();
  document.getElementById("Date").valueAsDate = new Date();
  date = getMDY(document.getElementById("Date").value);
  document.getElementById("Date").valueAsDate = new Date();
  var form = document.getElementById("aInfo");
  var list = document.getElementById("taskList");
  changePrettyDate();
  function handleKeys(e) {
    //Space key pressed
    if (e.keyCode == 32) {
      chrome.tabs.query({active: true, lastFocusedWindow: true}, tabs => {
        var url = tabs[0].url;
        document.getElementById("Link").value = url;
        // use `url` here inside the callback because it's asynchronous!
      });
    }
    //Backspace pressed
    if (e.keyCode == 8) {
      document.getElementById("Link").value = ' ';
    }
  }

  document.getElementById("Link").addEventListener("keydown", handleKeys);
  document.getElementById("Date").addEventListener('change', changePrettyDate);
  document.getElementById("Link").addEventListener('change', function () {
    if (isUrl(document.getElementById("Link").value)==false) {
      document.getElementById("LinkLabel").innerHTML = '';
      document.getElementById("FAVICON").setAttribute("src", '');
    }
    else {
      changeTitle();
    }
  });

  function clr() {
    chrome.notifications.clear('tst');
  }

  form.addEventListener('submit', function (event) {
    event.preventDefault();
    var name = document.getElementById("Name").value;
    var date = document.getElementById("Date").value;
    //Date formatting
    var month = parseInt(date.slice(5,7));
    var day = parseInt(date.slice(8,10));
    var year = parseInt(date.slice(0,4));
    var correctD = new Date(Date.UTC(year, month-1, day+1));
    var dateString = months[month] + " " + day + " ";
    var link = document.getElementById("Link").value;

    //Remove button formatting
    //Some slick string manipulation to get the link tag to work correctly
    if (name != '') {                                                                                                   //EDIT: turns out ids CAN have spaces, u just have to put quotes around the id. Well guess what too bad I am so not going to go through the entire thing just to not okokokokokokok
      if (isUrl(link)==false) {                                                                                         //inject date into id lol. ids cant have spaces so we replace the spaces with "ok", then we replace the oks with spaces again when we want to decode.
        list.innerHTML += '<li id=' + '"' + name + '">' + '<div class=item id=item nm=' + '"' + name + '">' + name +': ' + '<label class=dateStr id=' + String(correctD).replace(/ /g, 'ok') + '>' + dateString + '</label>' + '<strong>(<label class=daysTill>' + '</label>)</strong>' + '&nbsp' + '<button type=button class=Remove id=' + name + '>' + 'X' + '</button>' + '<div>' + '</li>';
        // chrome.storage.local.set({'tasklist': list.innerHTML});
        // localStorage.setItem('tasklist', list.innerHTML);
      }
      else {
      // localStorage.setItem('tasklist', list.innerHTML);
      // chrome.storage.local.set({'tasklist': list.innerHTML});
      list.innerHTML += '<li id=' + '"' + name + '">' + '<div class=item id=item nm=' + '"' + name + '">' + '<a target="_blank" href=' + link + '>' + name + '</a>' +': ' + '<label class=dateStr id=' + String(correctD).replace(/ /g, 'ok') + '>' + dateString + '</label>' + '<strong>(<label class=daysTill>' + '</label>)</strong>' + '&nbsp' + '<button type=button class=Remove id=' + name + '>' + 'X' + '</button>' + '<div>' + '</li>';
      // localStorage.setItem('tasklist', list.innerHTML);
      }
      document.getElementById("Name").value = "";
      document.getElementById("Link").value = "";
      document.getElementById("Link").focus();
    }
    // localStorage.setItem('tasklist', list.innerHTML);
    // chrome.storage.local.set({'tasklist': list.innerHTML});
    buttons = document.getElementsByClassName('Remove');
    numB = buttons.length;
    for (var i = 0; i < numB; i += 1) {
        buttons[i].addEventListener('click', function () {
          lists = document.getElementsByTagName('li');
          //target is the li element
          var target = this.parentElement.parentElement;
          target.style.opacity = '0';
          setTimeout(function(){target.remove(); localStorage.setItem('tasklist', list.innerHTML);}, 250);
//          this.parentElement.parentElement.remove();
          //localStorage.setItem('tasklist', list.innerHTML);
        });
    }
    tills = document.getElementsByClassName("daysTill");
    items = document.getElementsByClassName("item");
    dates = document.getElementsByClassName("dateStr");
    len = tills.length;

    for (var i = 0; i < len; i += 1) {
      var today = new Date();
      var originalID = dates[i].id;
      var dateInfo = originalID.replace(/ok/g, ' ');
      var storedDate = new Date(dateInfo);
      remainingDays = dateDiffInDays(today, storedDate);
      tills[i].innerHTML = remainingDays;
      tills[i].id = remainingDays;
      if (remainingDays == 1) {
        tills[i].style.color = "red";
      }

      else if (remainingDays == 0) {
        items[i].style.color = "DarkRed";
        items[i].style.fontWeight = 900;
        tills[i].innerHTML = 'TODAY';
        tills[i].style.color = "red";
      }

      else if (remainingDays < 0) {
        items[i].style.color = "red";
        items[i].style.fontWeight = 900;
        tills[i].innerHTML = 'OVERDUE';
        tills[i].style.color = "red";
      }

    }

    // chrome.storage.local.set({'tasklist': list.innerHTML});
    localStorage.setItem('tasklist', list.innerHTML);
  });

  // chrome.storage.local.get(['tasklist'], function(saved) {
  //   if (saved.tasklist) {
  //   	list.innerHTML = saved.tasklist;
  //   }
  //
  // });
  if (localStorage.getItem('tasklist')) {
  list.innerHTML = localStorage.getItem('tasklist');
  }

});
